src/libsyntax/codemap.rs

```diff
@@ -593,6 +609,32 @@ impl CodeMap {
         }
     }
 
+    /// Extend the given `Span` to just after the previous occurrence of `c`. Return the same span
+    /// if no character could be found or if an error occurred while retrieving the code snippet.
+    pub fn span_extend_to_prev_char(&self, sp: Span, c: char) -> Span {
+        if let Ok(prev_source) = self.span_to_prev_source(sp) {
+            let prev_source = prev_source.rsplit(c).nth(0).unwrap_or("").trim_left();
+            if !prev_source.is_empty() {
+                return sp.with_lo(BytePos(sp.lo().0 - prev_source.len() as u32));
+            }
+        }
+
+        sp
+    }
+
+    /// Extend the given `Span` to just after the previous occurrence of `pat`. Return the same span
+    /// if no character could be found or if an error occurred while retrieving the code snippet.
+    pub fn span_extend_to_prev_str(&self, sp: Span, pat: &str) -> Span {
```

Comment:
I think you could unify both of these methods writing it like this:

```rust
 pub fn span_extend_to_prev<P>(&self, sp: Span, pat: P) -> Span {
```

Reply:
I actually tried this version because I also thought that it would be sufficient to unify those methods, but I've got a compile error:
```rust
// Code
    pub fn span_extend_to_prev<P>(&self, sp: Span, pat: P) -> Span {
        if let Ok(prev_source) = self.span_to_prev_source(sp) {
            let prev_source = prev_source.rsplit(pat).nth(0).unwrap_or("").trim_left();
            if !prev_source.is_empty() {
                return sp.with_lo(BytePos(sp.lo().0 - prev_source.len() as u32));
            }
        }

        sp
    }

// Error
error[E0277]: the trait bound `P: std::ops::FnMut<(char,)>` is not satisfied
   --> libsyntax\codemap.rs:615:43
    |
615 |             let prev_source = prev_source.rsplit(pat).nth(0).unwrap_or("").trim_left();
    |                                           ^^^^^^ the trait `std::ops::FnMut<(char,)>` is not implemented for `P`
    |
    = help: consider adding a `where P: std::ops::FnMut<(char,)>` bound
    = note: required because of the requirements on the impl of `std::str::pattern::Pattern<'_>` for `P`

error[E0599]: no method named `nth` found for type `std::str::RSplit<'_, P>` in the current scope
   --> libsyntax\codemap.rs:615:55
    |
615 |             let prev_source = prev_source.rsplit(pat).nth(0).unwrap_or("").trim_left();
    |                                                       ^^^
    |
    = note: the method `nth` exists but the following trait bounds were not satisfied:
            `std::str::RSplit<'_, P> : std::iter::Iterator`

error: aborting due to 2 previous errors
```

I also tried being explicit on the type parameter by copying the bounds of `rsplit()`, but then I've got a lifetime error:
```rust
// Code
    pub fn span_extend_to_prev2<'a, P: Pattern<'a>>(&self, sp: Span, pat: P) -> Span 
        where <P as Pattern<'a>>::Searcher: ReverseSearcher<'a>
    {
        if let Ok(prev_source) = self.span_to_prev_source(sp) {
            let prev_source = prev_source.rsplit(pat).nth(0).unwrap_or("").trim_left();
            if !prev_source.is_empty() {
                return sp.with_lo(BytePos(sp.lo().0 - prev_source.len() as u32));
            }
        }

        sp
    }

// Error
error[E0597]: `prev_source` does not live long enough
   --> libsyntax\codemap.rs:628:31
    |
628 |             let prev_source = prev_source.rsplit(pat).nth(0).unwrap_or("").trim_left();
    |                               ^^^^^^^^^^^ borrowed value does not live long enough
...
632 |         }
    |         - borrowed value only lives until here
    |
note: borrowed value must be valid for the lifetime 'a as defined on the method body at 624:5...
   --> libsyntax\codemap.rs:624:5
    |
624 | /     pub fn span_extend_to_prev2<'a, P: Pattern<'a>>(&self, sp: Span, pat: P) -> Span
625 | |         where <P as Pattern<'a>>::Searcher: ReverseSearcher<'a>
626 | |     {
627 | |         if let Ok(prev_source) = self.span_to_prev_source(sp) {
...   |
634 | |         sp
635 | |     }
    | |_____^
```

Here are some links to the playground illustrating the errors: [first try][1], [second try][2].

Do you have an idea of how I could fix this?

[1]: https://play.rust-lang.org/?gist=f6797f5271da80c33d9498492f3c5c30&version=nightly
[2]: https://play.rust-lang.org/?gist=829569512050301a0df4057e10831039&version=nightly

Reply:
The first one is missing a trait bound, as written it would allow _any_ type to be passed as a pattern.

For the second, the string needs to outlive the method, which you can assure by [borrowing it instead](https://play.rust-lang.org/?gist=eb61a70a48ce013d254e24ce11aae74c&version=nightly).

Reply:
For the first one, I think I still have a C++ mental model where the template parameter is automatically bound by its use inside the function, so I thought this should work :).

Anyway for the second one, when I use the `Pattern<'a>` bound I'm not sure how to use a borrowed string instead of a `String` since I use the result of `span_to_prev_source()` which return `Result<String, SpanSnippetError>` (That's why I used `Option<String>` in the playground, to try to mirror this case). 

I could create a function like this:
```rust
fn span_and_source_extend_to_prev<'a, P: Pattern<'a>>(&self, sp: Span, prev_source: &'a str, pat: P) -> Span
    where <P as Pattern<'a>>::Searcher: ReverseSearcher<'a>
{
    let prev_source = prev_source.rsplit(pat).nth(0).unwrap_or("").trim_left();
    if !prev_source.is_empty() {
        sp.with_lo(BytePos(sp.lo().0 - prev_source.len() as u32))
    } else {
        sp
    }
}
```

But then I still need two more functions for `char` and `&str`, otherwise I run into the same problem: 
```rust
pub fn span_extend_to_prev_char(&self, sp: Span, c: char) -> Span {
    if let Ok(prev_source) = self.span_to_prev_source(sp) {
        self.span_and_source_extend_to_prev(sp, &prev_source, c)
    } else {
        sp
    }
}

pub fn span_extend_to_prev_str(&self, sp: Span, pat: &str) -> Span {
    if let Ok(prev_source) = self.span_to_prev_source(sp) {
        self.span_and_source_extend_to_prev(sp, &prev_source, pat)
    } else {
        sp
    }
}
```
So it seems to defeat the point.

What do you think?